certbot-compatibility-test/certbot_compatibility_test/configurators/common.py

```diff
@@ -37,25 +50,34 @@ def __init__(self, args):
         self.http_port = 80
         self.https_port = 443
         self._configurator: interfaces.Configurator
-        self._all_names = None
-        self._test_names = None
+        self._all_names: Optional[Set[str]] = None
+        self._test_names: Optional[Set[str]] = None
 
-    def has_more_configs(self):
+    def has_more_configs(self) -> bool:
         """Returns true if there are more configs to test"""
         return bool(self._configs)
 
     @abstractmethod
-    def cleanup_from_tests(self):
+    def cleanup_from_tests(self) -> None:
         """Performs any necessary cleanup from running plugin tests"""
 
-    def load_config(self):
+    def load_config(self) -> str:
         """Returns the next config directory to be tested"""
         shutil.rmtree(self.le_config.work_dir, ignore_errors=True)
         backup = os.path.join(self.le_config.work_dir, constants.BACKUP_DIR)
         os.makedirs(backup)
         return self._configs.pop()
 
-    def copy_certs_and_keys(self, cert_path, key_path, chain_path=None):
+    @overload
+    def copy_certs_and_keys(self, cert_path: str, key_path: str,
+                            chain_path: str) -> Tuple[str, str, str]: ...
+
+    @overload
+    def copy_certs_and_keys(self, cert_path: str, key_path: str,
+                            chain_path: Optional[str]) -> Tuple[str, str, Optional[str]]: ...
+
+    def copy_certs_and_keys(self, cert_path: str, key_path: str,
+                            chain_path: Optional[str] = None) -> Tuple[str, str, Optional[str]]:
```

Comment:
What is the purpose of doing this?

Reply:
It is an alternative of generic methods using `TypeVar` when the generic is a union of disjoint boundaries. Basically it allows to define several possible set of types for the method signature, both on the parameters types and the returned type.

Here it expresses the fact that the type of the third element in the returned tuple depends on the type of `chain_path`: `str` for a `str`, `Optional[str]` for an `Optional[str]`, because we know from the implementations that `chain_path` is returned as it in the third element of the tuple.

The two methods methods annotated with `@overload` define the signatures, then the actual implementation is given without annotation. Obviously the implementation can be typed and its boundary signature must be compatible with all signatures.

Reply:
Aha, that's cool. Thank you for the explanation.